### PR TITLE
feat: add cors support for api resource

### LINKED ISSUE #205: cors support for api resource
### Suggestion
Add cors middleware handling to the api resource. Enabled by default, with the ability to override headers.
It should also auto generate the options handlers per route.

```js
import { api } from '@nitric/sdk';

const mainApi = api('main', {
  cors: {
    allowOrigins: ['*'],
    allowMethods: ['PATCH', 'POST'],
    allowHeaders: ['*'],
    allowCredentials: true,
    maxAge: '300 seconds'
  },
});
```

### Value
Reduces the amount of boilerplate code in users apps.

### PR DESCRIPTION
**Pull request by davemooreuws**
closes #205 

Disabled by default

### CODE REVIEW

File: src/resources/api.ts [hunk 1]
```diff
@@ -276,6 +362,7 @@ export class Api<SecurityDefs extends string> extends Base<ApiDetails> {
     SecurityDefinition
   >;
   private readonly security?: Record<SecurityDefs, string[]>;
+  private readonly cors?: boolean | CorsOptions;
```
Comment (tjholm) [inline src/resources/api.ts:365]: I wonder if its worth implementing the CorsApi as an extension of the main API. Then we might be able to implement some TS shenanigans so we can remove methods from the main API interface in the case that CORS is enabled e.g. `Omit<Api, "options">`.

Comment (davemooreuws) [reply src/resources/api.ts:365]: Interesting point.

I think we should now consider the following:
- do we even need options as a public method? As they are used for preflight requests anyway.
- if we do allow them, would the users handler override our cors one.

File: src/resources/api.ts [hunk 2]
```diff
@@ -284,13 +371,23 @@ export class Api<SecurityDefs extends string> extends Base<ApiDetails> {
       path = '/',
       securityDefinitions = null,
       security = {} as Record<SecurityDefs, string[]>,
+      cors = true,
```
Comment (tjholm) [inline src/resources/api.ts:374]: I think we should leave this as undefined and assume false by default.

File: src/resources/api.ts (hunk 1 above)
Comment (davemooreuws) [reply src/resources/api.ts:375]: going to do it within the membrane, resolving this.

**Pull request closed without merging**